File: Engine/src/hook/player.cpp
/* 
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

#include "bridge/forwards/forward.h"
#include "global.h"
#include "manager/HookManager.h"
#include "module.h"
#include "vhook/hook.h"

#include "cstrike/component/PlayerPawnComponent.h"
#include "cstrike/entity/CBaseWeapon.h"
#include "cstrike/entity/PlayerController.h"
#include "cstrike/entity/PlayerPawn.h"
#include "cstrike/type/CNetworkGameServer.h"
#include "cstrike/type/CTakeDamageInfo.h"
#include "cstrike/type/CTakeDamageResult.h"

#include <safetyhook.hpp>

// #define PLAYER_EVENT_ASSERT
// #define WEAPON_EVENT_ASSERT

// forward to movement
extern float GetPlayerRunSpeedValue(CCSPlayerController* pController);

extern CCSPlayerPawn*       s_pMovementPawn;
extern CCSPlayerController* s_pMovementController;
extern CServerSideClient*   s_pMovementClient;

BeginMemberHookScope(CCSPlayerPawn)
{
    DeclareVirtualHook(Event_Killed, uint8_t, (CCSPlayerPawn * pPawn, CTakeDamageResult * pResult))
    {
        const auto pController = pPawn->GetController<CCSPlayerController*>();
        const auto pClient     = pController ? sv->GetClient(pController->GetPlayerSlot()) : nullptr;

        if (pClient)
        {
            g_pHookManager->Call_PlayerKilled(HookType_Pre, pPawn, pClient);
            forwards::OnPlayerKilledPre->Invoke(pClient, pController, pPawn, pResult);
        }

        const auto ret = Event_Killed(pPawn, pResult);

        if (pClient)
        {
            g_pHookManager->Call_PlayerKilled(HookType_Post, pPawn, pClient);
            forwards::OnPlayerKilledPost->Invoke(pClient, pController, pPawn, pResult);
        }

#ifdef PLAYER_EVENT_ASSERT
        LOG("CCSPlayerPawn::Event_Killed %p", pPawn);
#endif

        return ret;
    }

    DeclareVirtualHook(PlayerSpawn, void, (CCSPlayerPawn * pPawn))
    {
        const auto pController = pPawn->GetController<CCSPlayerController*>();
        const auto pClient     = pController ? sv->GetClient(pController->GetPlayerSlot()) : nullptr;

        if (pClient)
        {
            g_pHookManager->Call_PlayerSpawned(HookType_Pre, pPawn, pClient);
            forwards::OnPlayerSpawnPre->Invoke(pClient, pController, pPawn);
        }

        PlayerSpawn(pPawn);

        if (pClient)
        {
            g_pHookManager->Call_PlayerSpawned(HookType_Post, pPawn, pClient);
            forwards::OnPlayerSpawnPost->Invoke(pClient, pController, pPawn);
        }

#ifdef PLAYER_EVENT_ASSERT
        LOG("CCSPlayerPawn::PlayerSpawn %p", pPawn);
#endif
    }

    DeclareVirtualHook(PlayerPreThink, void, (CCSPlayerPawn * pPawn))
    {
        PlayerPreThink(pPawn);

        forwards::OnPlayerPreThink->Invoke(s_pMovementClient, s_pMovementController, pPawn);
#ifdef PLAYER_EVENT_ASSERT
        // LOG("CCSPlayerPawn::PlayerPreThink %p", pPawn);
#endif
    }

    DeclareVirtualHook(PlayerPostThink, void, (CCSPlayerPawn * pPawn))
    {
        PlayerPostThink(pPawn);

        forwards::OnPlayerPostThink->Invoke(s_pMovementClient, s_pMovementController, pPawn);
#ifdef PLAYER_EVENT_ASSERT
        // LOG("CCSPlayerPawn::PlayerPostThink %p", pPawn);
#endif
    }

#ifdef PLATFORM_WINDOWS
    static SafetyHookMid hk_ProcessMovement_AssignMaxSpeed{};

    static void ProcessMovement_AssignMaxSpeed(SafetyHookContext & ctx)
    {
        const auto pPawn       = s_pMovementPawn;
        const auto pController = s_pMovementController;
        const auto pClient     = s_pMovementClient;

        const float originalSpeed = ctx.xmm6.f32[0];
        auto        speed         = originalSpeed;
        const auto  action        = forwards::OnPlayerGetMaxSpeed->Invoke(pClient, pController, pPawn, originalSpeed, &speed);

        if (action == EHookAction::SkipCallReturnOverride)
        {
            ctx.xmm6.f32[0] = speed * GetPlayerRunSpeedValue(pController);
        }

        if (action == EHookAction::ChangeParamReturnDefault || action == EHookAction::ChangeParamReturnOverride || action == EHookAction::IgnoreParamReturnOverride)
        {
            FatalError("Not impl yet");
        }
    }
#else
    DeclareMemberDetourHook(GetPlayerMaxSpeed, float, (CCSPlayerPawn * pPawn))
    {
        const float originalSpeed = GetPlayerMaxSpeed(pPawn);

        const auto pController = s_pMovementController;
        const auto pClient     = s_pMovementClient;

        float      speed  = originalSpeed;
        const auto action = forwards::OnPlayerGetMaxSpeed->Invoke(pClient, pController, pPawn, originalSpeed, &speed);
        if (action == EHookAction::SkipCallReturnOverride)
        {
            return speed * GetPlayerRunSpeedValue(pController);
        }
        if (action == EHookAction::ChangeParamReturnDefault || action == EHookAction::ChangeParamReturnOverride || action == EHookAction::IgnoreParamReturnOverride)
        {
            FatalError("Not impl yet");
        }

        return originalSpeed * GetPlayerRunSpeedValue(pController);
    }
#endif
}

BeginMemberHookScope(CCSPlayer_WeaponServices)
{
    DeclareVirtualHook(SwitchWeapon, void, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon))
    {
        SwitchWeapon(pService, pWeapon);

        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();
        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
            forwards::OnPlayerSwitchWeapon->Invoke(sv->GetClient(pController->GetPlayerSlot()), pController, pPawn, pService, pWeapon);

#ifdef WEAPON_EVENT_ASSERT
        LOG("CCSPlayer_WeaponServices::SwitchWeapon %p", pWeapon);
#endif
    }

    DeclareMemberDetourHook(EquipWeapon, int64_t, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon))
    {
        const auto ret   = EquipWeapon(pService, pWeapon);
        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();
        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
            forwards::OnPlayerEquipWeapon->Invoke(sv->GetClient(pController->GetPlayerSlot()), pController, pPawn, pService, pWeapon, ret);

#ifdef WEAPON_EVENT_ASSERT
        LOG("CCSPlayer_WeaponServices::EquipWeapon %p", pWeapon);
#endif

        return ret;
    }

    DeclareVirtualHook(DropWeapon, void, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon, Vector * pVecTarget, Vector * pVelocity))
    {
        // 提前中断, 与游戏原本行为一致
        if (!pWeapon)
        {
            return;
        }

        DropWeapon(pService, pWeapon, pVecTarget, pVelocity);

        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();
        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
            forwards::OnPlayerDropWeapon->Invoke(sv->GetClient(pController->GetPlayerSlot()), pController, pPawn, pService, pWeapon);

#ifdef WEAPON_EVENT_ASSERT
        LOG("CCSPlayer_WeaponServices::DropWeapon %p", pWeapon);
#endif
    }

    DeclareVirtualHook(CanUse, bool, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon))
    {
        // 提前中断, 与游戏原本行为一致
        if (!pWeapon)
        {
            return false;
        }

#ifdef WEAPON_EVENT_ASSERT
        LOG("CCSPlayer_WeaponServices::CanUse %p", pWeapon);
#endif

        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();

        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
        {
            if (const auto pClient = sv->GetClient(pController->GetPlayerSlot()))
            {
                auto       result = true;
                const auto action = forwards::OnPlayerWeaponCanUse->Invoke(pClient, pController, pPawn, pWeapon, &result);
                if (action == EHookAction::SkipCallReturnOverride)
                {
                    return result;
                }

                if (action == EHookAction::IgnoreParamReturnOverride)
                {
                    CanUse(pService, pWeapon);
                    return result;
                }

                if (action == EHookAction::ChangeParamReturnDefault || action == EHookAction::ChangeParamReturnOverride)
                {
                    FatalError("Not impl yet");
                }
            }
        }

        return CanUse(pService, pWeapon);
    }

    DeclareMemberDetourHook(CanSwitch, bool, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon))
    {
        // 提前中断, 与游戏原本行为一致
        if (!pWeapon)
        {
            return false;
        }

        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();

        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
        {
            if (const auto pClient = sv->GetClient(pController->GetPlayerSlot()))
            {
                auto       result = true;
                const auto action = forwards::OnPlayerWeaponCanSwitch->Invoke(pClient, pController, pPawn, pWeapon, &result);
                if (action == EHookAction::SkipCallReturnOverride)
                {
                    return result;
                }

                if (action == EHookAction::ChangeParamReturnDefault || action == EHookAction::ChangeParamReturnOverride || action == EHookAction::IgnoreParamReturnOverride)
                {
                    FatalError("Not impl yet");
                }
            }
        }

        return CanSwitch(pService, pWeapon);
    }

    DeclareMemberDetourHook(CanEquip, bool, (CCSPlayer_WeaponServices * pService, CBaseWeapon * pWeapon))
    {
        const auto pPawn = pService->GetPawn<CCSPlayerPawn*>();

        if (const auto pController = pPawn->GetController<CCSPlayerController*>())
        {
            if (const auto pClient = sv->GetClient(pController->GetPlayerSlot()))
            {
                auto       result = true;
                const auto action = forwards::OnPlayerWeaponCanEquip->Invoke(pClient, pController, pPawn, pWeapon, &result);
                if (action == EHookAction::SkipCallReturnOverride)
                {
                    return result;
                }

                if (action == EHookAction::ChangeParamReturnDefault || action == EHookAction::ChangeParamReturnOverride || action == EHookAction::IgnoreParamReturnOverride)
                {
                    FatalError("Not impl yet");
                }
            }
        }

        return CanEquip(pService, pWeapon);
    }
}

void InstallPlayerHooks()
{
#ifdef PLATFORM_WINDOWS

    auto address = g_pGameData->GetAddress<void*>("CCSPlayer_MovementServices::ProcessMovement_AssignMaxSpeed");
    AssertPtr(address);

    auto result = safetyhook::MidHook::create(address, CCSPlayerPawn_Hooks::ProcessMovement_AssignMaxSpeed);
    if (!result.has_value())
    {
        FatalError("Failed to CCSPlayer_MovementServices::ProcessMovement_AssignMaxSpeed. ERROR: %s", g_szMidFuncHookErrors[result.error().type]);
    }

    CCSPlayerPawn_Hooks::hk_ProcessMovement_AssignMaxSpeed = std::move(*result);
    g_pHookManager->Register(&CCSPlayerPawn_Hooks::hk_ProcessMovement_AssignMaxSpeed);
#else
    InstallMemberDetourAutoSig(CCSPlayerPawn, GetPlayerMaxSpeed);
#endif

    InstallMemberDetourAutoSig(CCSPlayer_WeaponServices, EquipWeapon);
    InstallMemberDetourAutoSig(CCSPlayer_WeaponServices, CanSwitch);
    InstallMemberDetourAutoSig(CCSPlayer_WeaponServices, CanEquip);

    InstallVirtualHookAutoWithVTableAuto(CCSPlayer_WeaponServices, SwitchWeapon, server);
    InstallVirtualHookAutoWithVTableAuto(CCSPlayer_WeaponServices, DropWeapon, server);
    InstallVirtualHookAutoWithVTableAuto(CCSPlayer_WeaponServices, CanUse, server);

    InstallVirtualHookManualWithVTableAuto(CCSPlayerPawn, Event_Killed, server, "CBaseEntity::Event_Killed");
    InstallVirtualHookAutoWithVTableAuto(CCSPlayerPawn, PlayerSpawn, server);
    InstallVirtualHookAutoWithVTableAuto(CCSPlayerPawn, PlayerPreThink, server);
    InstallVirtualHookAutoWithVTableAuto(CCSPlayerPawn, PlayerPostThink, server);
}

File: Sharp.Core/Bridges/Forwards/Player.cs
/*
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Sharp.Shared.Enums;
using Sharp.Shared.Types;

namespace Sharp.Core.Bridges.Forwards;

internal static class Player
{
    public delegate void DelegatePlayerSpawn(nint ptrClient, nint ptrController, nint ptrPawn);

    public delegate void DelegatePlayerThink(nint ptrClient, nint ptrController, nint ptrPawn);

    public delegate void DelegatePlayerKilled(nint ptrClient, nint ptrController, nint ptrPawn, nint ptrResult);

    public delegate EHookAction DelegatePlayerRunCommandPre(nint ptrClient,
        nint                                                     ptrController,
        nint                                                     ptrPawn,
        nint                                                     ptrService,
        nint                                                     ptrUserCmd);

    public delegate void DelegatePlayerRunCommandPost(nint ptrClient,
        nint                                               ptrController,
        nint                                               ptrPawn,
        nint                                               ptrService,
        nint                                               ptrUserCmd,
        EHookAction                                        action);

    public delegate void DelegatePlayerProcessMove(nint ptrClient,
        nint                                            ptrController,
        nint                                            ptrPawn,
        nint                                            ptrService,
        nint                                            ptrMoveData);

    public delegate void DelegatePlayerWalkMove(nint ptrClient,
        nint                                         ptrController,
        nint                                         ptrPawn,
        nint                                         ptrService,
        nint                                         ptrMoveData,
        ref int                                      speed);

    public delegate void DelegateOnPlayerSwitchWeapon(nint ptrClient,
        nint                                               ptrController,
        nint                                               ptrPawn,
        nint                                               ptrService,
        nint                                               ptrWeapon);

    public delegate void DelegateOnPlayerEquipWeapon(nint ptrClient,
        nint                                              ptrController,
        nint                                              ptrPawn,
        nint                                              ptrService,
        nint                                              ptrWeapon,
        long                                              equipmentValue);

    public delegate void DelegateOnPlayerDropWeapon(nint ptrClient,
        nint                                             ptrController,
        nint                                             ptrPawn,
        nint                                             ptrService,
        nint                                             ptrWeapon);

    public delegate HookReturnValue<bool> DelegateOnPlayerWeaponCanUse(nint ptrClient,
        nint                                                                ptrController,
        nint                                                                ptrPawn,
        nint                                                                ptrWeapon);

    public delegate HookReturnValue<bool> DelegateOnPlayerWeaponCanSwitch(nint ptrClient,
        nint                                                                   ptrController,
        nint                                                                   ptrPawn,
        nint                                                                   ptrWeapon);

    public delegate HookReturnValue<bool> DelegateOnPlayerWeaponCanEquip(nint ptrClient,
        nint                                                                  ptrController,
        nint                                                                  ptrPawn,
        nint                                                                  ptrWeapon);

    public delegate HookReturnValue<float> DelegateOnPlayerGetMaxSpeed(nint ptrClient,
        nint                                                                ptrController,
        nint                                                                ptrPawn,
        float                                                               originalSpeed);

    public static event DelegatePlayerSpawn?             OnPlayerSpawnPre;
    public static event DelegatePlayerSpawn?             OnPlayerSpawnPost;
    public static event DelegatePlayerThink?             OnPlayerPreThink;
    public static event DelegatePlayerThink?             OnPlayerPostThink;
    public static event DelegatePlayerKilled?            OnPlayerKilledPre;
    public static event DelegatePlayerKilled?            OnPlayerKilledPost;
    public static event DelegatePlayerRunCommandPre?     OnPlayerRunCommandPre;
    public static event DelegatePlayerRunCommandPost?    OnPlayerRunCommandPost;
    public static event DelegatePlayerProcessMove?       OnPlayerProcessMovePre;
    public static event DelegatePlayerProcessMove?       OnPlayerProcessMovePost;
    public static event DelegatePlayerWalkMove?          OnPlayerWalkMove;
    public static event DelegateOnPlayerSwitchWeapon?    OnPlayerSwitchWeapon;
    public static event DelegateOnPlayerEquipWeapon?     OnPlayerEquipWeapon;
    public static event DelegateOnPlayerDropWeapon?      OnPlayerDropWeapon;
    public static event DelegateOnPlayerWeaponCanUse?    OnPlayerWeaponCanUse;
    public static event DelegateOnPlayerWeaponCanSwitch? OnPlayerWeaponCanSwitch;
    public static event DelegateOnPlayerWeaponCanEquip?  OnPlayerWeaponCanEquip;
    public static event DelegateOnPlayerGetMaxSpeed?     OnPlayerGetMaxSpeed;

    [UnmanagedCallersOnly]
    public static void OnPlayerSpawnPreExport(nint ptrClient, nint ptrController, nint ptrPawn)
        => OnPlayerSpawnPre?.Invoke(ptrClient, ptrController, ptrPawn);

    [UnmanagedCallersOnly]
    public static void OnPlayerSpawnPostExport(nint ptrClient, nint ptrController, nint ptrPawn)
        => OnPlayerSpawnPost?.Invoke(ptrClient, ptrController, ptrPawn);

    [UnmanagedCallersOnly]
    public static void OnPlayerPreThinkExport(nint ptrClient, nint ptrController, nint ptrPawn)
        => OnPlayerPreThink?.Invoke(ptrClient, ptrController, ptrPawn);

    [UnmanagedCallersOnly]
    public static void OnPlayerPostThinkExport(nint ptrClient, nint ptrController, nint ptrPawn)
        => OnPlayerPostThink?.Invoke(ptrClient, ptrController, ptrPawn);

    [UnmanagedCallersOnly]
    public static void OnPlayerKilledPreExport(nint ptrClient, nint ptrController, nint ptrPawn, nint ptrResult)
        => OnPlayerKilledPre?.Invoke(ptrClient, ptrController, ptrPawn, ptrResult);

    [UnmanagedCallersOnly]
    public static void OnPlayerKilledPostExport(nint ptrClient, nint ptrController, nint ptrPawn, nint ptrResult)
        => OnPlayerKilledPost?.Invoke(ptrClient, ptrController, ptrPawn, ptrResult);

    [UnmanagedCallersOnly]
    public static EHookAction OnPlayerRunCommandPreExport(nint ptrClient,
        nint                                                   ptrController,
        nint                                                   ptrPawn,
        nint                                                   ptrService,
        nint                                                   ptrUserCmd)
        => OnPlayerRunCommandPre?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrUserCmd) ?? EHookAction.Ignored;

    [UnmanagedCallersOnly]
    public static void OnPlayerRunCommandPostExport(nint ptrClient,
        nint                                             ptrController,
        nint                                             ptrPawn,
        nint                                             ptrService,
        nint                                             ptrUserCmd,
        EHookAction                                      action)
        => OnPlayerRunCommandPost?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrUserCmd, action);

    [UnmanagedCallersOnly]
    public static void OnPlayerProcessMovePreExport(nint ptrClient,
        nint                                             ptrController,
        nint                                             ptrPawn,
        nint                                             ptrService,
        nint                                             ptrMoveData)
        => OnPlayerProcessMovePre?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrMoveData);

    [UnmanagedCallersOnly]
    public static void OnPlayerProcessMovePostExport(nint ptrClient,
        nint                                              ptrController,
        nint                                              ptrPawn,
        nint                                              ptrService,
        nint                                              ptrMoveData)
        => OnPlayerProcessMovePost?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrMoveData);

    [UnmanagedCallersOnly]
    public static unsafe void OnPlayerWalkMoveExport(nint ptrClient,
        nint                                              ptrController,
        nint                                              ptrPawn,
        nint                                              ptrService,
        nint                                              ptrMoveData,
        int*                                              pSpeed)
    {
        if (OnPlayerWalkMove is null)
        {
            return;
        }

        ref var speed = ref Unsafe.AsRef<int>(pSpeed);
        OnPlayerWalkMove.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrMoveData, ref speed);
    }

    [UnmanagedCallersOnly]
    public static void OnPlayerSwitchWeaponExport(nint ptrClient,
        nint                                           ptrController,
        nint                                           ptrPawn,
        nint                                           ptrService,
        nint                                           ptrWeapon)
        => OnPlayerSwitchWeapon?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrWeapon);

    [UnmanagedCallersOnly]
    public static void OnPlayerEquipWeaponExport(nint ptrClient,
        nint                                          ptrController,
        nint                                          ptrPawn,
        nint                                          ptrService,
        nint                                          ptrWeapon,
        long                                          equipmentValue)
        => OnPlayerEquipWeapon?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrWeapon, equipmentValue);

    [UnmanagedCallersOnly]
    public static void OnPlayerDropWeaponExport(nint ptrClient,
        nint                                         ptrController,
        nint                                         ptrPawn,
        nint                                         ptrService,
        nint                                         ptrWeapon)
        => OnPlayerDropWeapon?.Invoke(ptrClient, ptrController, ptrPawn, ptrService, ptrWeapon);

    [UnmanagedCallersOnly]
    public static unsafe EHookAction OnPlayerWeaponCanUseExport(nint ptrClient,
        nint                                                         ptrController,
        nint                                                         ptrPawn,
        nint                                                         ptrWeapon,
        bool*                                                        ptrReturn)
    {
        if (OnPlayerWeaponCanUse is null)
        {
            return EHookAction.Ignored;
        }

        var action = OnPlayerWeaponCanUse.Invoke(ptrClient, ptrController, ptrPawn, ptrWeapon);

        *ptrReturn = action.ReturnValue;

        return action.Action;
    }

    [UnmanagedCallersOnly]
    public static unsafe EHookAction OnPlayerWeaponCanSwitchExport(nint ptrClient,
        nint                                                            ptrController,
        nint                                                            ptrPawn,
        nint                                                            ptrWeapon,
        bool*                                                           ptrReturn)
    {
        if (OnPlayerWeaponCanSwitch is null)
        {
            return EHookAction.Ignored;
        }

        var action = OnPlayerWeaponCanSwitch.Invoke(ptrClient, ptrController, ptrPawn, ptrWeapon);

        *ptrReturn = action.ReturnValue;

        return action.Action;
    }

    [UnmanagedCallersOnly]
    public static unsafe EHookAction OnPlayerWeaponCanEquipExport(nint ptrClient,
        nint                                                           ptrController,
        nint                                                           ptrPawn,
        nint                                                           ptrWeapon,
        bool*                                                          ptrReturn)
    {
        if (OnPlayerWeaponCanEquip is null)
        {
            return EHookAction.Ignored;
        }

        var action = OnPlayerWeaponCanEquip.Invoke(ptrClient, ptrController, ptrPawn, ptrWeapon);

        *ptrReturn = action.ReturnValue;

        return action.Action;
    }

    [UnmanagedCallersOnly]
    public static unsafe EHookAction OnPlayerGetMaxSpeedExport(nint ptrClient,
        nint                                                        ptrController,
        nint                                                        ptrPawn,
        float                                                       originalSpeed,
        float*                                                      ptrReturn)
    {
        if (OnPlayerGetMaxSpeed is null)
        {
            return EHookAction.Ignored;
        }

        var action = OnPlayerGetMaxSpeed.Invoke(ptrClient, ptrController, ptrPawn, originalSpeed);

        *ptrReturn = action.ReturnValue;

        return action.Action;
    }
}
